### PR TITLE
Advanced forms: conditional relationships in case index

### PR DESCRIPTION
**Pull request by orangejenny**
Opening for review. I'm still working on tests.

Spec: https://docs.google.com/document/d/1oVZWmFd2xG9TmNe-gi-rgY0bZGOYbD0p7Z0Hl_WUmHo/

screen shot 2018-12-20 at 5 29 41 pm

@millerdev or maybe @proteusvacuum

### CODE REVIEW

**Comment by orangejenny** [comment]
Fixed old tests, added new tests.

File: corehq/apps/app_manager/tests/test_extension_case.py [hunk 1]
```diff
@@ -227,46 +242,32 @@ def test_xform_case_block_index_valid_relationship(self):
...
-        """ Owner id should be automatically set if there are any non-extension indices"""
+    def test_advanced_xform_autoset_owner_id(self):
+        """
+            Owner id should be automatically set if there are any non-extension indices.
+            It should never be set if there are any dynamically-specified indices.
+        """
+
+        def _test_relationships(relationships, expected):
+            advanced_open_action = AdvancedOpenCaseAction.wrap({'case_indices': [{
+                'tag': 'tag{}'.format(int(random.random() * 10000)),
```
Comment (millerdev) [inline corehq/apps/app_manager/tests/test_extension_case.py]: Is it important for test passage that this tag value is unique? If so using a random value could result in a rare intermittently failing test if two of the random values picked turned out to be the same. Might be worth rearranging things a bit:
```suggestion
                'tag': 'tag{}'.format(i),
```

File: corehq/apps/app_manager/tests/test_extension_case.py [hunk 2]
```diff
@@ -227,46 +242,32 @@ def test_xform_case_block_index_valid_relationship(self):
...
+                'reference_id': 'case',
+                'relationship': r,
+            } for r in relationships]})
```
Comment (millerdev) [inline corehq/apps/app_manager/tests/test_extension_case.py]: This is also required if the suggestion above is accepted.
```suggestion
            } for i, r in enumerate(relationships)]})
```

**Review by millerdev — APPROVED** [review]

File: corehq/apps/app_manager/tests/test_extension_case.py (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/app_manager/tests/test_extension_case.py]: Good point, will update in a followup PR.

**Comment by proteusvacuum** [comment]
What does this do to the app summary? And exports? 
I think there are a lot of assumptions baked into those things about the different types of relationship cases can have. 
Mostly worried about:
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/export/models/new.py#L1823-L1826

I think app summary is mostly handled gracefully (i.e. cases with this would show up as having a "question" relationship):

**Comment by orangejenny** [comment]
@proteusvacuum I have not tested those, thanks for the heads up.

App summary looks okay, seems that it doesn't display the relationship but instead the reference ID:
screen shot 2018-12-21 at 11 46 53 am

I'll look into exports.

**Pull request merged**